### PR TITLE
Use es-index-cleaner golang implementation

### PR DESCRIPTION
**Pull request by pavolloffay**
Signed-off-by: Pavol Loffay <p.loffay@gmail.com>

Updates #3179

### CODE REVIEW

File: cmd/es-index-cleaner/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+ARG base_image
```
Comment (jpkrohling) [inline cmd/es-index-cleaner/Dockerfile:1]: @Ashmita152, are you interested in reviewing this?

**Review by jpkrohling — COMMENTED** [review]
LGTM, just want to hear from @Ashmita152 on the Dockerfile changes.

**Comment by pavolloffay** [comment]
The index cleaer integration test is failing bc it cannot build the image via buildx. The buildx uses different CI runner.

```
CGO_ENABLED=0 installsuffix=cgo go build -trimpath -o ./cmd/es-index-cleaner/es-index-cleaner-linux-amd64 ./cmd/es-index-cleaner/main.go
make[1]: Leaving directory '/home/runner/work/jaeger/jaeger'
docker build --target release -t jaegertracing/jaeger-es-index-cleaner:latest --build-arg base_image=localhost:5000/baseimg_alpine:latest  --build-arg TARGETARCH=amd64 cmd/es-index-cleaner
Sending build context to Docker daemon  10.05MB

Step 1/5 : ARG base_image
Step 2/5 : FROM $base_image AS release
Get http://localhost:5000/v2/: dial tcp [::1]:5000: connect: connection refused
make: *** [Makefile:330: docker-images-elastic] Error 1
```

**Comment by Ashmita152** [comment]
@pavolloffay I think you would need to start registry container in elasticsearch integration tests i.e. add this section https://github.com/jaegertracing/jaeger/blob/master/.github/workflows/ci-hotrod.yml#L12-L16 in https://github.com/jaegertracing/jaeger/blob/master/.github/workflows/ci-elasticsearch.yml for fixing the CI failures.

File: Makefile [hunk 2]
```diff
@@ -115,10 +115,10 @@ grpc-plugin-storage-integration-test:
 .PHONY: test-compile-es-scripts
 test-compile-es-scripts:
 	docker run --rm -v ${PWD}:/tmp/jaeger python:3-alpine3.11 /usr/local/bin/python -m py_compile /tmp/jaeger/plugin/storage/es/esRollover.py
-	docker run --rm -v ${PWD}:/tmp/jaeger python:3-alpine3.11 /usr/local/bin/python -m py_compile /tmp/jaeger/plugin/storage/es/esCleaner.py
 
 .PHONY: index-cleaner-integration-test
-index-cleaner-integration-test: docker-images-elastic
+index-cleaner-integration-test:
+	$(MAKE) BASE_IMAGE=alpine:3.13 docker-images-elastic
```
Comment (Ashmita152) [inline Makefile:121]: Curious why we are overwriting BASE_IMAGE here ? I would like to keep a single source of truth for all base images and debug images.

File: cmd/es-index-cleaner/app/index_filter.go [hunk 3]
```diff
@@ -45,15 +45,12 @@ func (i *IndexFilter) filter(indices []Index) []Index {
 	if !i.Rollover && !i.Archive {
 		// daily indices
 		reg, _ = regexp.Compile(fmt.Sprintf("^%sjaeger-(span|service|dependencies)-\\d{4}%s\\d{2}%s\\d{2}", i.IndexPrefix, i.IndexDateSeparator, i.IndexDateSeparator))
-	} else if !i.Rollover && i.Archive {
-		// daily archive
-		reg, _ = regexp.Compile(fmt.Sprintf("^%sjaeger-span-archive", i.IndexPrefix))
+	} else if i.Archive {
+		// archive works only for rollover
+		reg, _ = regexp.Compile(fmt.Sprintf("^%sjaeger-span-archive-\\d{6}", i.IndexPrefix))
```
Comment (Ashmita152) [inline cmd/es-index-cleaner/app/index_filter.go:50]: Personal opinion, I think we can rewrite the if else conditional into:

```
if i.Archive {
    // archive works only for rollover
    reg, _ = regexp.Compile(fmt.Sprintf("^%sjaeger-span-archive-\\d{6}", i.IndexPrefix))    
} else if i.Rollover {
    // rollover
    reg, _ = regexp.Compile(fmt.Sprintf("^%sjaeger-(span|service)-\\d{6}", i.IndexPrefix))
} else {
    // daily indices
    reg, _ = regexp.Compile(fmt.Sprintf("^%sjaeger-(span|service|dependencies)-\\d{4}%s\\d{2}%s\\d{2}", i.IndexPrefix, i.IndexDateSeparator, i.IndexDateSeparator))
}
```

File: Makefile (hunk 2 above)
Comment (pavolloffay) [reply Makefile:121]: I don't like that building the images requires running docker registry at localhost. This complicates local and CI setup.

Comment (Ashmita152) [reply Makefile:121]: I think reverting this change back should fix the current CI failure.

 Here it is trying to tag the image name as alpine:3.13
```
make BASE_IMAGE=alpine:3.13 docker-images-elastic
make[1]: Entering directory '/home/runner/work/jaeger/jaeger'
docker buildx build -t alpine:3.13 --push \
	--build-arg root_image=alpine:3.13 \
	--build-arg cert_image=alpine:3.13 \
	--platform=linux/amd64,linux/s390x,linux/ppc64le,linux/arm64 \
	docker/base
```
which it later tries to push in the official alpine image resulting in
```
error: failed to solve: authorization status: 401: authorization failed
```

Comment (pavolloffay) [reply Makefile]: yes this was the problem. I will try to play with tit and see if I could run it locally.

Can you run the full build locally? I had some issues with running the registry locally.

**Comment by pavolloffay** [comment]
the CI is all green now

File: Makefile (hunk 2 above)
Comment (albertteoh) [reply Makefile]: @pavolloffay I had issues with local builds as well in the past; I managed to get it going I think. Maybe this helps? https://github.com/jaegertracing/jaeger/pull/2948/files#r666710826

**Comment by pavolloffay** [comment]
PR rebased and ready for review/merge.

File: Makefile [hunk 4]
```diff
@@ -316,10 +320,13 @@ docker-images-cassandra:
 	docker build -t $(DOCKER_NAMESPACE)/jaeger-cassandra-schema:${DOCKER_TAG} plugin/storage/cassandra/
 	@echo "Finished building jaeger-cassandra-schema =============="
 
+docker-images-elastic: TARGET = release
```
Comment (albertteoh) [inline Makefile:323]: I thought `--target` would be useful if a Dockerfile contained multiple build stages like `debug` and `release`. However, from what I see, there is only a `release` build stage in cmd/es-index-cleaner/Dockerfile. Would it be useful to still include the target param in this case?

Comment (pavolloffay) [reply Makefile]: True, this can be removed. Done.

**Review by albertteoh — APPROVED** [review]
lgtm

**Pull request merged**